### ISSUE TITLE
Using "site" query parameter results in "Unable to open this type of link"

### ISSUE THREAD

**Issue by GuiltyDolphin** [duckduckgo/Android#347]
### App version

5.9.1 (50901) (via F-Droid)

### Device make and model

OnePlus 5

### Android version

8.1.0

### Steps to reproduce the bug

1. Into the address bar type `foo site:duckduckgo.com` (or any query + any site)

2. Press enter (try to load page)

### Expected behavior

A DuckDuckGo search opens with the query

### Actual behavior

A toast pops up briefly with the phrase "Unable to open this type of link"

**Comment by dineshkv** [comment]
I believe this is a feature request.

**Comment by GuiltyDolphin** [comment]
@dineshkv I think this is a bug (i.e., something broke) - I noticed it because I used to do it and it worked (it definitely works in version 5.7.5), but now no longer works

**Comment by dineshkv** [comment]
Oh okay. If I want to work on this are PRs going to be accepted?

**Comment by CDRussell** [comment]
Thanks for reporting the `site:` problem; that was an unintended regression of another recent change.

Happy to accept a PR for this if someone wants to pick it up. And thanks in advance, if so!

**Comment by CDRussell** [comment]
The problem is likely to be found in `SpecialUrlDetector`. We recently switched things around to better handle custom app schemes, and in the process broke queries like this that look like a URL scheme but should just be treated as a normal search query.

https://github.com/duckduckgo/Android/blob/develop/app/src/main/java/com/duckduckgo/app/browser/SpecialUrlDetector.kt#L42

**Comment by In2deep73** [comment]
I have some fixes to propose if @dineshkv decides not to work on this.

**Comment by dineshkv** [comment]
@In2deep73 please go ahead as I have not started working on it yet.

**Comment by CDRussell** [comment]
Additional test case to cover under this fix: searches like `std::fabs`

**Comment by subsymbolic** [comment]
Thanks to everyone who has submitted their PRs and ideas here, it has been really valuable. Taking all the ideas into consideration I would like to move forward with separating the logic for items that have an authority from those that do not when detecting the type in our `SpecialUrlDetector`. I imagine splitting the logic into two categories as follows:

**Category 1: Items with a Uri.authority component. This covers web links and intent URIs**
The behavior for these would be unchanged. Examples:
- http://duckduckgo.com - Should continue to be classed as `UrlType.WebUrl`
- twitter://user?14504859 - Should continue to be classed as  `UrlType.IntentType` or `UrlType.Unknown` if twitter is not installed
- unknown://23432 - Should continue to be classed as `UrlType.Unknown`

**Category 2: Items without a Uri.authority component. This covers special schemes and search queries**
These are items that do not have an authority. Previously these defaulted to `UrlType.Unknown`, they should now default to `UrlType.SearchQuery`. Examples:
- `tel:1234567` - tel is one of our accepted schemes so should continue to be classed as `UrlType.Telephone`
- `about:default` - We should change this from `UrlType.Unknown` to a new type `UrlType.About` as we do know about this and should be explicit about it. We should then add an entry for the new type in `BrowserWebViewClient:shouldOverride(webView: WebView, url: Uri)` and handle it as we do `UrlType.Unknown`.
- `site:duckduckgo.com search text` - site is not one of our accepted special schemes so should default to `UrlType.SearchQuery`
- `search text site:duckduckgo.com` - Again `UrlType.SearchQuery`
- `search text unknown:item` - Again `UrlType.SearchQuery`
- `search text` - Again `UrlType.SearchQuery`

To summarize my suggestion: start by identifying if the input has an authority (category 1 above) or not (category 2 above) and then allocating a type from there. It looks like @donfuxx is already on the case (thanks!) in https://github.com/duckduckgo/Android/pull/503. I'll leave the task with him and if this isn't solved soon, we'll bring it in-house.

Note: fixing ip addresses / ports isn't considered here but I have noted that that's a pain point too so we'll get that cleaned up as part of a separate task after this one.

**Comment by subsymbolic** [comment]
Thanks for all the activity here. I have accepted one PR that goes most of the way towards fixing this, however there are still a number of edge cases. While `foo site:duckduckgo.com` now works
`site:duckduckgo.com foo` and `site:duckduckgo.com` do not.

It is quite tricky to fix this without breaking other links so this will require some wider refactoring. We have logged the problem internally and will clean it up soon. Thanks again for bringing it to our attention.

Internal link: https://app.asana.com/0/361428290920652/1126831098940688

**Comment by slavgirl** [comment]
This issue is occurring currently.  Is anyone attending this? Can someone direct me to a site with instructions for getting a patch, if there is one?  I'm totally NOT a coder nor familiar with language, just trying to find a remedy.

Thank you!!

**Comment by cmonfortep** [comment]
Hello @GuiltyDolphin and the rest of followers here. After internal discussion with the team, we've agreed that there's no plan in the foreseeable future to do changes in how our address bar handles concrete keyboards or patterns as we do on our SERP page. Should this change in the future, we'll update this conversation appropriately.

For now, closing this issue as the main complaint was fixed some time ago.

**Comment by rolinger** [comment]
@cmonfortep - based on your last comment, if this is in regards to deeplinking urlschemes, then I am going to have to add  specific JS code to my website that determines if the browser is "DuckDuckGo" that the user will then need to rescan the QR Code and choose to open the link with Chrome/Safari/Edge/FF - because your browser won't open deeplinks.

Issue I created regarding this is referenced here: https://github.com/duckduckgo/Android/issues/1458

**Comment by cmonfortep** [comment]
@rolinger we will investigate your problem in the issue you've just created. This issue is related with queries manually submitted in our Url bar. Your scenario may be different.
